src/core/providers/memory/qgsmemoryprovider.cpp

```diff
@@ -556,28 +557,39 @@ bool QgsMemoryProvider::deleteFeatures( const QgsFeatureIds &id )
 
 bool QgsMemoryProvider::addAttributes( const QList<QgsField> &attributes )
 {
-  for ( QList<QgsField>::const_iterator it = attributes.begin(); it != attributes.end(); ++it )
+  for ( QgsField field : attributes )
   {
-    switch ( it->type() )
+    if ( !supportedType( field ) )
+      continue;
+
+    // Make sure added attributes typeName correspond to a native type name
+    bool isNativeTypeName = false;
+    NativeType nativeTypeCandidate( QString(), QString(), QVariant::Invalid );
+    for ( const NativeType &nativeType : nativeTypes() )
     {
-      case QVariant::Int:
-      case QVariant::Double:
-      case QVariant::String:
-      case QVariant::Date:
-      case QVariant::Time:
-      case QVariant::DateTime:
-      case QVariant::LongLong:
-      case QVariant::StringList:
-      case QVariant::List:
-      case QVariant::Bool:
-      case QVariant::ByteArray:
+      if ( nativeType.mTypeName.toLower() == field.typeName().toLower() )
+      {
+        isNativeTypeName = true;
         break;
-      default:
-        QgsDebugMsg( "Field type not supported: " + it->typeName() );
+      }
+
+      if ( nativeType.mType == field.type()
+           && nativeTypeCandidate.mType == QVariant::Invalid )
+        nativeTypeCandidate = nativeType;
+    }
+    if ( !isNativeTypeName )
+    {
+      if ( nativeTypeCandidate.mType == QVariant::Invalid )
+      {
+        QgsDebugMsg( "Field type not supported: " + field.typeName() );
         continue;
+      }
+
+      field.setTypeName( nativeTypeCandidate.mTypeName );
```

Comment:
I am not sure about this. If a field is added from another provider, and the instances of fields are shared, this would modify the field type name in the original provider. Should I do a deep copy of the field before?

Reply:
It's not necessary -- by design all the implicitly shared classes in QGIS (like QgsField) are copy-on-write, so the deep copy will happen automatically in the background when you call a setter like this.

Reply:
Ok thank you!